runtime/compiler/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,245 @@
+   static int32_t recompileThreshold = comp->getOptions()->getJProfilingLoopRecompThreshold();
+   if (trace())
+      traceMsg(comp, "Loop Recompilation Base Threshold = %d\n",recompileThreshold);
+   // Iterating backwards to avoid losing original block associated with the test location tree tops in case we have found multiple
+   // recompilation test location in same block.
+   for (auto testLocationIter = testLocations.rbegin(), testLocationEnd = testLocations.rend(); testLocationIter != testLocationEnd; ++testLocationIter)
+      {
+      TR::TreeTop *asyncCheckTreeTop = testLocationIter->first.first;
+      TR::Block *originalBlock = testLocationIter->first.second;
+      TR::Node *node = asyncCheckTreeTop->getNode();
+      int32_t depth = testLocationIter->second;
+      if (trace())
+         traceMsg(comp, "block_%d, n%dn, depth = %d\n",originalBlock->getNumber(), asyncCheckTreeTop->getNode()->getGlobalIndex(), depth);
+      TR_ByteCodeInfo bci = asyncCheckTreeTop->getNode()->getByteCodeInfo();
+      
+      TR::Node *root = bfi->generateBlockRawCountCalculationSubTree(comp, node, trace());
+      
+      // If we got a bad counters/ bad block frequency info, above API would return NULL, if that is the case we can not generate a recompilation test for that location.
+      if (!root)
+         {
+         TR::DebugCounter::incStaticDebugCounter(comp, TR::DebugCounter::debugCounterName(comp, "jprofiling.instrument/badcounters/(%s)", comp->signature()));
+         continue;
+         }
+      dumpOptDetails(comp, "%s Add recompilation test after asyncCheck node n%dn\n", optDetailString(), node->getGlobalIndex());
+
+      //Splitting the original block from AsyncCheckTT
+      TR::Block *remainingCodeBlock = originalBlock->split(asyncCheckTreeTop->getNextTreeTop(), cfg, true, true);
+      TR::Block *callRecompileBlock = TR::Block::createEmptyBlock(node, comp, UNKNOWN_COLD_BLOCK_COUNT);
+      callRecompileBlock->setIsCold(true);
+      
+      // jitRetranslateCallerWithPrep Helper call
+      TR::TreeTop *callTree = TR::TransformUtil::generateRetranslateCallerWithPrepTrees(node, TR_PersistentMethodInfo::RecompDueToJProfiling, comp);
+      callTree->getNode()->setIsProfilingCode();
+      callRecompileBlock->append(callTree);
+      cfg->addNode(callRecompileBlock);
+      const char * const name = TR::DebugCounter::debugCounterName(comp,
+      "recompilationHelper/(%s)/%d",
+      comp->signature(),depth);
+      TR::DebugCounter::prependDebugCounter(comp, name, callTree);
+      
+      // Code to calculate the raw frequency of loop from block counters and comparing with the adjusted recompilation threshold.
+      // threshold for this particular test location is calculated from the base recompile threshold and the nesting depth of the loop
+      // Putting a higher threshold limit to 10K currently to prohibit running profiling body for too long.
+      int32_t threshold = recompileThreshold << ((depth-1)*1);
```

Comment:
why the *1? Should there be something to stop overflow if depth gets too deep - unlikely, but possible...